src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,940 @@
+   end
+   if isdefined(G, :gens) && isdefined(H, :gens)
+      if G.gens==H.gens return true  end       #TODO this works only if the == between elements does not keep track of the parent
+   end
+   return G.X==H.X
+end
+
+
+# this saves the value of x.X
+# x_gap = x.X if this is already known, x_gap = nothing otherwise
+function lies_in(x::MatElem, G::MatrixGroup, x_gap)
+   if base_ring(x)!=G.ring || nrows(x)!=G.deg return false, x_gap end
+   if isone(x) return true, x_gap end
+   if isdefined(G,:gens)
+      for g in G.gens
+         if x==g.elm
+            return true, x_gap
+         end
+      end
+   end
+   if isdefined(G,:descr) && G.descr==:GL
+      if det(x)!=0
+         return true, x_gap
+      else
+         return false, x_gap
+      end
+   elseif isdefined(G,:descr) && G.descr==:SL
+      if det(x)==1
+         return true, x_gap
+      else
+         return false, x_gap
+      end
+   else
+      if x_gap==nothing x_gap = G.mat_iso(x) end
+     # x_gap !=nothing || x_gap = G.mat_iso(x)
+      return GAP.Globals.in(x_gap,G.X), x_gap
+   end
+end
+
+Base.in(x::MatElem, G::MatrixGroup) = lies_in(x,G,nothing)[1]
+
+function Base.in(x::MatrixGroupElem, G::MatrixGroup)
+   if isdefined(x,:X) return lies_in(x.elm,G,x.X)[1]
+   else return lies_in(x.elm,G,nothing)[1]
+   end
+end
+
+# embedding an element of type MatElem into a group G
+function (G::MatrixGroup)(x::MatElem)
+   vero, x_gap = lies_in(x,G,nothing)
+   vero || throw(ArgumentError("Element not in the group"))
+   if x_gap==nothing return MatrixGroupElem(G,x)
+   else return MatrixGroupElem(G,x,x_gap)
+   end
+end
+
+# embedding an element of type MatrixGroupElem into a group G
+function (G::MatrixGroup)(x::MatrixGroupElem)
+   if isdefined(x,:X)
+      if isdefined(x,:elm)
+         vero = lies_in(x.elm,G,x.X)[1]
+         vero || throw(ArgumentError("Element not in the group"))
+         return MatrixGroupElem(G,x.elm,x.X)
+      else
+         GAP.Globals.in(x.X, G.X) || throw(ArgumentError("Element not in the group"))
+         return MatrixGroupElem(G,x.X)
+      end
+   else
+      vero, x_gap = lies_in(x.elm,G,nothing)
+      vero || throw(ArgumentError("Element not in the group"))
+      if x_gap==nothing return MatrixGroupElem(G,x.elm)
+      else return MatrixGroupElem(G,x.elm,x_gap)
+      end
+   end
+end
+
+# embedding a nxn array into a group G
+function (G::MatrixGroup)(L::AbstractVector)
+   x = matrix(G.ring, G.deg, G.deg, L)
+   return G(x)
+end
+
+########################################################################
+#
+# Methods on elements
+#
+########################################################################
+
+# TODO: we are not currently keeping track of the parent
+function ==(x::MatrixGroupElem{S,T},y::MatrixGroupElem{S,T}) where {S,T}
+   if isdefined(x,:X) && isdefined(y,:X) return x.X==y.X
+   else return x.elm==y.elm
+   end
+end
+
+# TODO: we wish to multiply / conjugate also matrices with different parents ?
```

Comment:
A situation occurring quite often where we multiply matrices with different parents is when one of the parents is a subgroup of the other one; in such a case, the ideal output parent would be the bigger group, but this could lead to many useless and expensive `issubgroup` and membership tests.
Anyway, I added a parameter `check` to the embedding functions. The user can disable this check if he is sure of the membership and skip all the checks.
A possibility is to set the parent of a product as simply `GL(n,q)`.

Reply:
Magma, for example, sets the parent of `x*y` as `GL(n,q)` whenever `x` and `y` have different parents (but same degree and base ring, of course). This is for sure the easiest way to proceed and avoid plenty of useless checks and membership tests.